client/rpcserver/websocket.go

```diff
@@ -0,0 +1,235 @@
+	"decred.org/dcrdex/dex"
+	"decred.org/dcrdex/dex/msgjson"
+	"decred.org/dcrdex/dex/ws"
+)
+
+var (
+	// Time allowed to read the next pong message from the peer. The
+	// default is intended for production, but leaving as a var instead of const
+	// to facilitate testing.
+	pongWait = 60 * time.Second
+	// Send pings to peer with this period. Must be less than pongWait. The
+	// default is intended for production, but leaving as a var instead of const
+	// to facilitate testing.
+	pingPeriod = (pongWait * 9) / 10
+	// A client id counter.
+	cidCounter int32
+	unbip      = dex.BipIDSymbol
+)
+
+type wsClient struct {
+	*ws.WSLink
+	mtx  sync.Mutex
+	cid  int32
+	quit func()
+}
+
+func newWSClient(ip string, conn ws.Connection, hndlr func(msg *msgjson.Message) *msgjson.Error) *wsClient {
+	return &wsClient{
+		WSLink: ws.NewWSLink(ip, conn, pingPeriod, hndlr),
+		cid:    atomic.AddInt32(&cidCounter, 1),
+	}
+}
+
+// handleWS handles the websocket connection request, creating a ws.Connection
+// and a websocketHandler thread.
+func (s *RPCServer) handleWS(w http.ResponseWriter, r *http.Request) {
+	// If the IP address includes a port, remove it.
+	ip := r.RemoteAddr
+	// If a host:port can be parsed, the IP is only the host portion.
+	host, _, err := net.SplitHostPort(ip)
+	if err == nil && host != "" {
+		ip = host
+	}
+	wsConn, err := ws.NewConnection(w, r, pingPeriod+pongWait)
+	if err != nil {
+		log.Errorf("ws connection error: %v", err)
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+	go s.websocketHandler(wsConn, ip)
+}
+
+// websocketHandler handles a new websocket client by creating a new wsClient,
+// starting it, and blocking until the connection closes. This method should be
+// run as a goroutine.
+func (s *RPCServer) websocketHandler(conn ws.Connection, ip string) {
+	log.Debugf("New websocket client %s", ip)
+	// Create a new websocket client to handle the new websocket connection
+	// and wait for it to shutdown.  Once it has shutdown (and hence
+	// disconnected), remove it.
+	var cl *wsClient
+	cl = newWSClient(ip, conn, func(msg *msgjson.Message) *msgjson.Error {
+		return s.handleMessage(cl, msg)
+	})
+	s.mtx.Lock()
+	s.clients[cl.cid] = cl
+	s.mtx.Unlock()
+	defer func() {
+		cl.mtx.Lock()
+		if cl.quit != nil {
+			cl.quit()
+		}
+		delete(s.clients, cl.cid)
+		cl.mtx.Unlock()
+	}()
+	cl.Start()
+	cl.WaitForShutdown()
+	log.Tracef("Disconnected websocket client %s", ip)
+}
+
+// handleMessage handles the websocket message, calling the right handler for
+// the route.
+func (s *RPCServer) handleMessage(conn *wsClient, msg *msgjson.Message) *msgjson.Error {
+	log.Tracef("message of type %d received for route %s", msg.Type, msg.Route)
+	if msg.Type == msgjson.Request {
+		handler, found := wsHandlers[msg.Route]
+		if !found {
+			// If a this request exits in routes, call it.
+			if _, found := routes[msg.Route]; found {
+				handler = wsHandleRequest
+			}
```

Comment:
perhaps

```go
} else {
	return msgjson.NewError(msgjson.RPCUnknownRoute, "unknown route '"+msg.Route+"'")
}
```

So you don't have to check `found` twice if you got it out of `wsHandlers` right away.